File: Data Structure & Algorithm/Template/Binary Search/Find Upper bound/Binary Search.go
package main

import "fmt"

// Find Upper bound
func binarySearch(nums []int, target int) int {
	left := 0
	right := len(nums) - 1

	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] <= target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	if left-1 >= 0 && nums[left-1] == target {
		return left - 1
	}

	return -1
}

func main() {
	nums := []int{-1, 0, 3, 5, 9, 12}
	target := 9
	res := binarySearch(nums, target)
	fmt.Println(res)
}

File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Models/Binary Search/Find Upper bound/Binary Search.py
from typing import List


# Find Upper bound
class Solution:

    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] <= target:
                left = mid + 1
            else:
                right = mid - 1

        if left - 1 >= 0 and nums[left - 1] == target:
            return left - 1

        return -1


# test
solution = Solution()
res = solution.search([-1, 0, 3, 5, 9, 12], 9)
print(res)
